### PR TITLE
[test] Add riscv compliance testing support

### PR DESCRIPTION
**Pull request by tjaychen**

### CODE REVIEW

**Comment by tjaychen** [comment]
continuing from #495

**Comment by tjaychen** [comment]
@imphil 
As i understand it, what I should do is the following:
1. vendor in riscv_compliance (current PR)
2. upload OpenTitan patch to lowrisc/riscv_compliance (this [PR](https://github.com/lowRISC/riscv-compliance/pull/2))
3. Once 2 is merged, updated current PR with patch information from patch repo.
4. Look into merging OpenTitan into the main riscv-compliance repo.

Is that what you had in mind?

**Comment by asb** [comment]
From my point of view, I think the value of vendoring in a version of the riscv-compliance suite that doesn't run on earl grey is limited. As such, it makes sense to vendor in a version patched for earl grey version and for us to review that. If that PR also has a commit that adds execution of the compliance suite to the CI, that would make it even easier to review.

**Comment by tjaychen** [comment]
> From my point of view, I think the value of vendoring in a version of the riscv-compliance suite that doesn't run on earl grey is limited. As such, it makes sense to vendor in a version patched for earl grey version and for us to review that. If that PR also has a commit that adds execution of the compliance suite to the CI, that would make it even easier to review.

@asb, that's fair.  Then I should wait for this [PR](https://github.com/lowRISC/riscv-compliance/pull/2), and I'll update this one accordingly.  That way whatever gets vendor'd in will run in top_earlgrey directly.  

I was thinking the CI addition should probably be a separate PR.

**Comment by asb** [comment]
Exactly where to split PRs is always a trade-off. If it's a pretty simple change to add to CI (which also means reviewers won't be distracted by in-depth review of that change) it can be pretty helpful, as it provides a shared demonstration that it works as expected and continues to do so through any updates.

**Comment by tjaychen** [comment]
@asb can you have a look? 
I'm guessing this is what you wanted? I've applied the patch along with the vendoring, so everything should work now.  I've also added a CI target, but I am removing tests that I know will fail.

The main thing that's missing is the patch_repo.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -259,6 +259,33 @@ jobs:
...
     displayName: 'Upload partial distribution artifacts'
 
+- job: "riscv_compliance_tests"
+  displayName: "Execute RISC-V compliance tests"
+  pool: "Default"
+  dependsOn:
+    - top_earlgrey_verilator
+  steps:
+  - bash: |
+      sudo apt-get install -y python3 python3-pip build-essential srecord python3-setuptools zlib1g-dev libusb-1.0 libftdi1-dev libftdi1-2 libssl-dev \
```
Comment (tjaychen) [inline azure-pipelines.yml]: eventually i'm going to include fpga riscv compliance testing, so maybe that means i would need to build the spiflash tools? if they eventually get built earlier and archived, i'll remove some of the dependencies here.

File: sw/vendor/riscv_compliance.vendor.hjson [hunk 2]
```diff
@@ -13,8 +14,17 @@
   exclude_from_upstream: [
     "doc",
     "spec",
-    "riscv-target",
     "riscv-ovpsim",
+    "riscv-target/Codasip-simulator",
```
Comment (tjaychen) [inline sw/vendor/riscv_compliance.vendor.hjson]: originally i removed riscv-target, and hoped that the patching would bring it back for the opentitan directory. That didn't work, so I manually called out each sub-directory.

File: sw/vendor/riscv_compliance.vendor.hjson [hunk 3]
```diff
@@ -13,8 +14,17 @@
   exclude_from_upstream: [
     "doc",
```
Comment (tjaychen) [inline sw/vendor/riscv_compliance.vendor.hjson]: all the docs / specs are removed since they caused CI failures with hugo.

File: test/systemtest/conftest.py [hunk 4]
```diff
@@ -133,6 +133,14 @@ def logfile(pytestconfig):
     log = pytestconfig.getoption('log')
     if not log:
         return
-    path = Path(log).resolve()
-    assert not path.is_dir()
+    # The strict option is only availabe on pythong 3.6
```
Comment (asb) [inline test/systemtest/conftest.py]: pythong -> python

File: test/systemtest/conftest.py [hunk 5]
```diff
@@ -133,6 +133,14 @@ def logfile(pytestconfig):
...
+    # CI currently uses 3.5
```
Comment (asb) [inline test/systemtest/conftest.py]: This is true, but more accurate to say that "However, Python 3.5.2 is our minimum supported version."

I think this should be pulled out to separate patch and could even be pulled out to a separate PR as it's a worthwhile fix even without the riscv compliance context.

**Comment by imphil** [comment]
Can we get https://github.com/lowRISC/riscv-compliance/pull/2 merged first, and then return to this PR to include it into our tree and enable it in CI?

**Comment by tjaychen** [comment]
> Can we get [lowRISC/riscv-compliance#2](https://github.com/lowRISC/riscv-compliance/pull/2) merged first, and then return to this PR to include it into our tree and enable it in CI?

btw @imphil , in case it's not clear.  I'm actually making edits in the other PR.  Then I in parallel patch it over here (because I need to test it anyways).  So as I'm updating there, I will also keep it in sync here, with the understanding this PR will not go in until the other is merged.

File: azure-pipelines.yml [hunk 6]
```diff
@@ -290,6 +290,35 @@ jobs:
       ci/run_verilator_pytest.sh
     displayName: 'Execute tests'
 
+- job: "riscv_compliance_tests"
+  displayName: "Execute RISC-V compliance tests"
+  pool: "Default"
+  dependsOn:
+    - top_earlgrey_verilator
```
Comment (tjaychen) [inline azure-pipelines.yml]: @mcy I am not sure if this is dependency is 100% necessary.
When I did not have it earlier, I think one of the jobs actually failed, and I just assumed it was a verilator conflict over the pty.

I'm actually not sure how isolated these jobs are supposed to be..

File: sw/vendor/riscv_compliance.vendor.hjson [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+{
+  name: "riscv_compliance",
+  target_dir: "riscv_compliance",
+
+  upstream: {
+    url: "https://github.com/riscv/riscv-compliance.git",
+    rev: "master",
+  },
```
Comment (imphil) [inline sw/vendor/riscv_compliance.vendor.hjson]: Now that you've merged your changes into the `ot` branch of https://github.com/lowRISC/riscv-compliance, please add `patch_repo` section like this:

```hjson
  patch_repo: {
    url: "https://github.com/lowRISC/riscv-compliance.git",
    rev_base: "master",
    rev_patched: "ot",
  },
```

You can then re-run the vendor script with ```--refresh-patches --commit --verbose` to generate a commit with the patches and the upstream sources. 

This should replace the `[vendor] Vendor in riscv compliance code ` and `[test] Add riscv_compliance patch` commits of this PR.

File: sw/vendor/README.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# Work with software code in external repositories
+
+Please refer to [vendor_hw](../../doc/vendor_hw.md) for more details on rationale of vendoring external repositories
+
+
+## RISCV_COMPLIANCE
```
Comment (imphil) [inline sw/vendor/README.md]: RISC-V Compliance Suite

File: sw/vendor/README.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+RISCV_COMPLIANCE defines a series of tests that to ensure the riscv core is compliant.
```
Comment (imphil) [inline sw/vendor/README.md]: RISCV_COMPLIANCE -> "The RISC-V Compliance Suite"
riscv core -> RISC-V system
compliant -> complies with the RISC-V Specification.

File: sw/vendor/README.md [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+Normally such a test would be run at the core level instead of the system level, but doing so confers advantages to ensure system integration did not alter the original properties of the core (for example breaking byte or half-word accesses)
```
Comment (imphil) [inline sw/vendor/README.md]: nit: add `.` at the end.

File: sw/vendor/README.md [hunk 11]
```diff
@@ -0,0 +1,14 @@
+# Work with software code in external repositories
+
+Please refer to [vendor_hw](../../doc/vendor_hw.md) for more details on rationale of vendoring external repositories
```
Comment (imphil) [inline sw/vendor/README.md]: nit: add `.` at end of sentence.

File: sw/vendor/README.md [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+
+## RISCV_COMPLIANCE
+
+RISCV_COMPLIANCE defines a series of tests that to ensure the riscv core is compliant.
+Normally such a test would be run at the core level instead of the system level, but doing so confers advantages to ensure system integration did not alter the original properties of the core (for example breaking byte or half-word accesses)
+
+To vendor in riscv compliance, please use teh commands below
```
Comment (imphil) [inline sw/vendor/README.md]: riscv compliance -> the RISC-V Compliance Suite
teh -> the
nit: add . at the end.

File: test/systemtest/conftest.py [hunk 13]
```diff
@@ -133,6 +133,14 @@ def logfile(pytestconfig):
...
+    # The strict option is only availabe on python 3.6
+    # CI currently uses >=3.5.2
+    # Turns out however even not using resolve doesn't work.
+    # The logging function in 3.5 uses os.path.isabs to check whether
+    # path is absolute and does not accept POSIXPATH objects
+    # path = Path(log).resolve(strict=False)
+    # assert not path.is_dir()
+
+    path = os.path.abspath(log)
+    assert not os.path.isdir(path)
```
Comment (imphil) [inline test/systemtest/conftest.py]: This should be a separate PR.

File: sw/vendor/riscv_compliance.vendor.hjson (hunk 7 above)
Comment (tjaychen) [reply sw/vendor/riscv_compliance.vendor.hjson]: yep will do this today. I just didnt get a chance to finish this part when I took off.

File: test/systemtest/conftest.py (hunk 13 above)
Comment (tjaychen) [reply test/systemtest/conftest.py]: could we keep it as part of this one? It's already a separate commit within the PR.

Comment (tjaychen) [reply test/systemtest/conftest.py]: o oops, yeah this should at leasst be another commit.  Let me fix it.

File: sw/vendor/README.md (hunk 11 above)
Comment (tjaychen) [reply sw/vendor/README.md]: fixed.

File: sw/vendor/README.md (hunk 8 above)
Comment (tjaychen) [reply sw/vendor/README.md]: fixed.

File: sw/vendor/README.md (hunk 9 above)
Comment (tjaychen) [reply sw/vendor/README.md]: fixed.

File: sw/vendor/README.md (hunk 10 above)
Comment (tjaychen) [reply sw/vendor/README.md]: fixed.

File: sw/vendor/README.md (hunk 12 above)
Comment (tjaychen) [reply sw/vendor/README.md]: fixed.

File: sw/vendor/riscv_compliance.vendor.hjson (hunk 7 above)
Comment (tjaychen) [reply sw/vendor/riscv_compliance.vendor.hjson]: for this portion, can i do it as a separate PR? It seems like the mainline riscv_compliance has updated, so the patches in the patch repo do not apply correctly anymore.

I'm assuming what I need to do now is first update the lowrisc port to tip of tree, make my changes on top of that, and then try the vendor flow again.  Is that right?

Or alternatively if I could modify the vendor script to checkout a specific version instead of always using head of tree.

Comment (imphil) [reply sw/vendor/riscv_compliance.vendor.hjson]: you can and should push directly to our fork repository (no need for PRs there). Rebase `ot` on top of the upstream master, and then push the upstream master and ot both to the forked repository. See also "How to manage patches in a Git repository" in https://docs.opentitan.org/doc/ug/vendor_hw/, or let me know if you're lost and I'll do it.

Comment (tjaychen) [reply sw/vendor/riscv_compliance.vendor.hjson]: i have been reading that, but i guess i'm not sure what you mean by push directly (sorry not a git expert). 

Let me give it a try later, right now I need to figure out why CI isn't working again even though it worked last time.

Comment (tjaychen) [reply sw/vendor/riscv_compliance.vendor.hjson]: alright..so i think i did that..but it seems with the latest updates, riscv_compliance is now failing some tests.

I really prefer to get this one in for now.  Then we can go back and debug the issues and also include the patch repo.  Is that okay with you?

File: azure-pipelines.yml [hunk 14]
```diff
@@ -219,6 +219,47 @@ jobs:
...
+      export VERILATED_SYSTEM_PATH="$BIN_DIR/hw/top_earlgrey/Vtop_earlgrey_verilator"
+      ci/run_verilator_pytest.sh
+    displayName: 'Execute tests'
+
+- job: "riscv_compliance_tests"
+  displayName: "Execute RISC-V compliance tests"
+  pool: "Default"
+  dependsOn:
+    - top_earlgrey_verilator
+    - execute_verilated_tests
```
Comment (imphil) [inline azure-pipelines.yml]: That dependency is unnecessary, isn't it?

File: azure-pipelines.yml (hunk 6 above)
Comment (imphil) [reply azure-pipelines.yml]: This dependency is necessary, as you need to build the simulation first.

File: azure-pipelines.yml [hunk 15]
```diff
@@ -219,6 +219,47 @@ jobs:
...
+        && sudo pip3 install -U -r python-requirements.txt
+    displayName: 'Install dependencies'
+  - template: 'ci/download-artifacts-template.yml'
+  - bash: |
+      . util/build_consts.sh
+      export TARGET_SIM="$BIN_DIR/hw/top_earlgrey/Vtop_earlgrey_verilator"
+      export RISCV_DEVICE=rv32imc
+      export RISCV_TARGET=opentitan
+      export OT_TARGET=verilator
+      export OT_TOOLS="${TOOLCHAIN_PATH}/bin"
```
Comment (imphil) [inline azure-pipelines.yml]: Please choose one style: either `${BIN_DIR}` above, or `$TOOLCHAIN_PATH` here.

File: azure-pipelines.yml [hunk 16]
```diff
@@ -219,6 +219,47 @@ jobs:
...
+      cd sw/vendor/riscv_compliance
+      make RISCV_ISA=rv32i
```
Comment (imphil) [inline azure-pipelines.yml]: You're intentionally not executing the other variants? If so, please add a comment. (https://github.com/lowRISC/ibex/blob/master/azure-pipelines.yml#L123)

File: azure-pipelines.yml [hunk 17]
```diff
@@ -284,6 +308,7 @@ jobs:
     - sw_build
     - top_earlgrey_verilator
     - top_earlgrey_nexysvideo
+    - riscv_compliance_tests
```
Comment (imphil) [inline azure-pipelines.yml]: That dependency is unnecessary.

File: azure-pipelines.yml (hunk 14 above)
Comment (tjaychen) [reply azure-pipelines.yml]: so i actually put this here on purpose because i did not want the riscv tests to run if the normal verilated tests fail (since those are super short and fast, and the riscv tests take a bit longer).

As i'm saying that now though, I'm guessing this does not really work since the pass / failure of a particular dependency probably is not factored in...

File: azure-pipelines.yml (hunk 17 above)
Comment (tjaychen) [reply azure-pipelines.yml]: my original rationale was the same as above..i was thinking that we wouldn't want to do a release unless the testing phases all worked (i was actually wondering why the normal verilator tests werent there either).

I'll take it for now, but do let me know what you think.

File: azure-pipelines.yml (hunk 15 above)
Comment (tjaychen) [reply azure-pipelines.yml]: ah good catch, thanks.

File: azure-pipelines.yml (hunk 16 above)
Comment (tjaychen) [reply azure-pipelines.yml]: comment added.

Comment (tjaychen) [reply azure-pipelines.yml]: i'm going to follow on this PR with another round of debugging on why the latest fail, and also enable the other variants.

File: azure-pipelines.yml (hunk 14 above)
Comment (tjaychen) [reply azure-pipelines.yml]: i added a condition check instead of taking it out. Let me know what you think, or if this is the wrong rationale, happy to take it out also.

Comment (imphil) [reply azure-pipelines.yml]: Creating a dependency like that serializes steps and increases end-to-end CI latency; that's bad for developer productivity: you want to get results as quickly as possible.

Also, think of this scenario: you have compliance suite and the other tests fail for different reasons. On the first push, you see that the verilator tests fail, and fix that. Now you push again, and see that the compliance tests fail. You need to do another push to fix that. As result, you've lost ~20 min waiting for another CI run, and even more in terms of productivity as you have to context-switch between tasks.

So the general rule: only introduce dependencies if you need them for correctness. It's much cheaper (at this level) to buy machine time than developer time.

File: azure-pipelines.yml (hunk 17 above)
Comment (imphil) [reply azure-pipelines.yml]: Dependencies are only needed for correctness. The overall pass/fail status of the pipeline isn't influenced by that.

In this case, we want to upload artifacts as soon as we can. You want to be able to use the CI-generated artifacts and do local debugging on them, e.g. to figure out why a test in CI failed.

On the point of releasing only binaries which have passed CI: we currently do that by first running CI on a specific commit on master, and then tagging that commit. That tagging will re-run CI and, assuming the output is deterministic, we get a working release. That's not ideal, but good enough and actually pretty common. If you want to get things fully correct, you'd want to "promote" release candidate builds to releases after they have passed testing. That requires a couple tricks to get the version in the binaries and the git tagging correct. Let's save that for another day.

File: azure-pipelines.yml (hunk 14 above)
Comment (tjaychen) [reply azure-pipelines.yml]: i see what you mean, but i think it's pretty unlikely (in this specific scenario) for our sanity tests to fail and somehow riscv CI to pass though... and since the riscv suite is fairly big (likely to exceed 70 tests long term, each compiled separately and loaded), it will definitely take a ton of time.  My view was that it'd be unnecessary to run it unless we know for sure the system functions at least at a basic level.

I'll take out the dependency for now though.

File: azure-pipelines.yml (hunk 17 above)
Comment (tjaychen) [reply azure-pipelines.yml]: o actually, i did not know we could download the CI generated artifacts for local debugging. I thought we could only download the released versions, am I missing something?

In any case, for this case, I've removed the dependency.

File: azure-pipelines.yml (hunk 14 above)
Comment (tjaychen) [reply azure-pipelines.yml]: @imphil can you have a look now?

File: azure-pipelines.yml (hunk 17 above)
Comment (imphil) [reply azure-pipelines.yml]: If you click on any Azure build, e.g. https://dev.azure.com/lowrisc/opentitan/_build/results?buildId=6361&view=results, click on the top-right Artifacts button, and download all outputs that we produce. The opentitan-dist one is the file that will be uploaded to GH releases if we do a tagged release.

File: azure-pipelines.yml (hunk 6 above)
Comment (tjaychen) [reply azure-pipelines.yml]: sorry i was actually referring to the execute* dependency, not the build dependency.

File: azure-pipelines.yml (hunk 17 above)
Comment (tjaychen) [reply azure-pipelines.yml]: thanks @imphil !

**Comment by tjaychen** [comment]
I just wanted to thank @imphil for his endless patience in reviewing this PR and providing guidance!
